### PR TITLE
Bug 1329524 - What changed at timestamp UI

### PR DESCRIPTION
**Pull request by andela-hngerebara**
UI to show all history logs.
**Things done:**

- Fields can be deleted if not required. Then each deleted field is then available in the add filter drop-down.

- Filtering can be done for multiple usernames or emails.

- The end date begins from the start date so you cannot select an end date that is before the start date.

- Based on the tabs selected and the associated checkboxes, the various endpoints are called.

- Paginated the results displayed

screen shot 2017-10-25 at 5 53 56 pm

screen shot 2017-10-25 at 5 53 15 pm

- Filtering can be done based on the date range and username specified. 
screen shot 2017-10-25 at 5 50 28 pm

- Created a modal that contains the details of each object

screen shot 2017-10-25 at 5 50 57 pm

### CODE REVIEW

**Comment by andela-hngerebara** [comment]
What I have so far calls all endpoints based on values entered and it makes me have a lot of conditions checked. So I was thinking a better approach should be to have 3 endpoints based on the tabs clicked rather than having just one endpoint. Each tab should consume one endpoint having queries passed to be filtered and the result returned.  I feel that will be a lot easier to maintain as opposed to having just one endpoint.

**Comment by mozbhearsum** [comment]
When you say "tabs", are you talking about the Change Type checkboxes?

**Comment by andela-hngerebara** [comment]
@mozbhearsum 
RPR, Scheduled Changes, and Required signoffs are tabs. Each tab has filter options which include separate checkboxes. 
- RPR has (Rules, Releases, and permissions)
- Scheduled Changes has (Rules, Releases, Permissions, Permissions required signoff, Product required signoff)
- Required Signoff has (Permissions required signoff, Product required signoff).

File: ui/app/js/controllers/history_data_controller.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+angular.module("app").controller("HistoryDataCtrl",
+function($scope, $http, $modalInstance, hs, hs_ct) {
+  $scope.hs = hs;
+  $scope.hs_ct = hs_ct;
+  $scope.rules_modal= false;
+  $scope.cancel = function () {
+    $modalInstance.dismiss('cancel');
+  };
+  // $scope.checkChangeType = function() {
```
Comment (collin5) [inline ui/app/js/controllers/history_data_controller.js]: I think we should completely remove this code since we are not using it.

Comment (andela-hngerebara) [reply ui/app/js/controllers/history_data_controller.js]: Thanks @collin5 . I intend to use it later though. It is still a [WIP] but I'll clean up when I'm done

**Comment by andela-hngerebara** [comment]
@mozbhearsum Thanks for the feedback. I will make these changes today and update my PR.

**Comment by andela-hngerebara** [comment]
@mozbhearsum So sorry. I hadn't updated the branch when you checked it out. You can try the filters now.

**Comment by mozbhearsum** [comment]
Wow, this is looking great! It's very functional and quite fast. A few smaller things to improve:
* The "Product/Channel" column should only be "Product" for Releases and Permissions Required Signoffs (neither of them have a channel).
* The view button (the eyeball) would be better in the first column -- where the primary keys are.
* When available, Rules should show Alias instead of Rule ID (some Rules will have one, others won't)
* In the Releases view, it would be really, really good to show the actual data changed along with the other columns. You can see an example of how to do this by going to http://localhost:8080/releases/Firefox-mozilla-central-nightly-latest and clicking "View Diff".
* We need a link in the navbar for this awesome page, so people can find it! Maybe "History" or "Changelog" on the far right side?

**Comment by andela-hngerebara** [comment]
@mozbhearsum All done.

**Comment by mozbhearsum** [comment]
Can you sync this up with #430?

**Comment by mozbhearsum** [comment]
#430 is merged now!

**Comment by andela-hngerebara** [comment]
@mozbhearsum I have already fixed all conflicts and the front end tests all pass locally. The backend tests still fail and I can't find the failing tests locally. I tried pushing the code so I can find the failing tests via the live logs but both the front end tests and back-end tests now fail. This is the error I now get. "image library/balrogtest: not found"
It is a docker error and now I can't really find the failing tests to fix.

**Comment by mozbhearsum** [comment]
It looks like those failures will be fixed by #455. If you merge master into your branch and then push it, the tests should rerun.

**Comment by andela-hngerebara** [comment]
@mozbhearsum All done

**Comment by mozbhearsum** [comment]
@andela-hngerebara, it looks like your branch may still be out of date against master. Eg: this branch still has the /all_rules/history endpoint, whereas master doesn't. Let me know if you need a hand getting up to date.

**Comment by andela-hngerebara** [comment]
Thanks, @mozbhearsum.  The branch is in sync with the master. I added 5 new endpoints for all objects required such that each endpoint retrieves data from the existing ones so filtering can be done easily.
E.g  "all_rules/history' returns { "Rules": {}, "Rules scheduled change": {}}
"all_releases" returns { "Release": {}, "Release scheduled change": {}}

File: ui/app/js/services/history_service.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+angular.module("app").factory('History', function($http, $q, ScheduledChanges, Helpers) {
+    var service = {
+      getHistory: function(filterParams, page){
+        var url = '/api/' +(filterParams.objectValue) +'/history?';
+        url += '&page=' + page;
```
[the remark this answers is not in the document]
Comment (andela-hngerebara) [reply ui/app/js/services/history_service.js]: Okay

File: ui/app/templates/history.html [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            </pagination>
+          </span>
+          <span style="float: right" class="count">{{history_count || 0}} item(s) found </span>
+        </div>
+
+        <table class="table table-bordered table-hover table-striped">
+          <thead>
+            <tr>
+              <th>sno</th>
+              <th ng-show="revision.rule_id">Rule Id / Alais</th>
```
[the remark this answers is not in the document]
Comment (andela-hngerebara) [reply ui/app/templates/history.html]: Okay I'll fix it

File: ui/app/js/controllers/history_controller.js [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    }
+    
+    searchHistoryLimit = 10;
+    combinedResults = [];
+
+    $scope.searchHistory = function() {
+      $scope.searchResult.length = 0;
+      $scope.loading = true;
+      checkParameters();
+      $scope.$watch("currentPage", function(page) {
```
[the remark this answers is not in the document]
Comment (andela-hngerebara) [reply ui/app/js/controllers/history_controller.js]: Okay

File: ui/app/js/controllers/history_controller.js [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+      
+    //Date range functions- To disable dates before start date
+    function startDateOnSetTime() {
+      $scope.$broadcast("start-date-changed");
+    }
+    function endDateOnSetTime() {
+      $scope.$broadcast("end-date-changed");
+    }
+
+    function startDateBeforeRender($dates) {
```
[the remark this answers is not in the document]
Comment (andela-hngerebara) [reply ui/app/js/controllers/history_controller.js]: Thanks

File: auslib/web/admin/views/scheduled_changes.py [hunk 6]
```diff
@@ -207,8 +207,10 @@ def _get_filters_all(self, obj):
         where = [getattr(self.history_table, f) == query.get(f) for f in query]
         where.append(self.history_table.data_version != null())
         request = connexion.request
-        if hasattr(self.history_table, 'product'):
-            where.append(self.history_table.product != null())
+        if request.args.get('product'):
+            where.append(self.history_table.base_product == request.args.get('product'))
+        if request.args.get('channel'):
+            where.append(self.history_table.base_channel == request.args.get('channel'))
```
[the remark this answers is not in the document]
Comment (andela-hngerebara) [reply auslib/web/admin/views/scheduled_changes.py]: Okay

**Comment by andela-hngerebara** [comment]
@mozbhearsum I have fixed the issues raised above.

**Comment by mozbhearsum** [comment]
I just started looking at this and it doesn't appear to work anymore. For example, requests such as http://localhost:8080/api/rules/history?page=1&changed_by=ffxbld&timestamp_from=1517461200000&timestamp_to=1519227000000&product=Firefox&channel=beta don't return any results - but there should be plenty of history in that time range...

**Comment by andela-hngerebara** [comment]
@mozbhearsum 
The search is done using the exact values.
Please confirm that changed_by=ffxbld has history from 1 February to 21st February 2018 based on the timestamp you used. From my end, ffxbld does not have any history with the date range you specified.
I just tried this and it works fine

http://localhost:8080/api/rules/history?page=1&changed_by=balrog-ffxbld&timestamp_from=1514828043000&timestamp_to=1516556043000&product=Firefox&channel=beta

http://localhost:8080/api/rules/history?page=1&changed_by=ffxbld&timestamp_from=1496668103000&timestamp_to=1516556043000&product=Firefox&channel=beta

**Comment by mozbhearsum** [comment]
Ah, my bad! I didn't realize your branch didn't have the new production dump path in it, so I wasn't looking far enough back in history to find anything!

I'll have another look at this this week.

File: auslib/web/common/history_all.py [hunk 7]
```diff
@@ -47,55 +54,73 @@ def _get_histories(table, obj, process_revisions_callback=None):
...
     """GET /rules/history"""
     history_table = dbo.rules.history
-    return _get_histories(history_table, get_rules)
-
-
-def get_releases_history():
+    rules = _get_histories(history_table, get_rules)
+    history = {
+        'rules': rules,
+        'sc_rules': RuleScheduledChangeHistoryView().get_all()
```
Comment (mozbhearsum) [inline auslib/web/common/history_all.py]: It looks like we're not apply any filters to this, which means we get unexpected results sometimes, eg: https://screenshots.firefox.com/zEoaUuLSSF7RyqXe/localhost

Let's fix this in a follow-up.

File: auslib/web/common/history_all.py [hunk 8]
```diff
@@ -47,55 +54,73 @@ def _get_histories(table, obj, process_revisions_callback=None):
...
     """GET /releases/history"""
     history_table = dbo.releases.history
-    return _get_histories(history_table, get_releases, process_release_revisions)
-
-
-def get_permissions_history():
+    releases = _get_histories(history_table, get_releases, process_release_revisions)
+    releases_history = {
+        'releases': releases,
+        'sc_releases': ReleaseScheduledChangeHistoryView().get_all()
```
Comment (mozbhearsum) [inline auslib/web/common/history_all.py]: Same thing here.

**Pull request merged**